## Main
This directory contains the scripts for differential accessibility analysis to identify cell-type or cluster-specific marker cCREs and TFs. It also contains scripts for differential analysis between MDD and neurotypical controls and downstream interpretation.

## To identify snATAC-seq chromatin cis-regulatory regions (cCREs)

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/snATAC_cCREs)

## To generate peak-count matrices using snATACs-seq cCREs

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/peak_count_matrices)

## To perform differential chromatin analysis between MDD and neurotypical controls

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/MDD_differential_accessibility)

## To perform differential transcription factor (TF) motif analysis between clusters and MDD and controls

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/differential_motifs)

## To create cluster-specific TF gene-regulatory networks 

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/TF_gene_regulatory_networks)

## To perform enrichment of DLPFC eQTLs/sQTLs in DLPFC cCREs

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/eQTL_sQTL_enrich_in_peaks)

## To perform spatial and functional characterization of microglia

[scripts](https://github.com/MGSSdouglas/snATAC_MDD/tree/main/3_differential_analysis/microglia_profiling)

